server/services/invidious.ts

```diff
@@ -0,0 +1,416 @@
+          codecs: r.codecs,
+        }));
+        log.info(
+          `DASH manifest summary url=${manifestUrl} total=${reps.length} variants=${JSON.stringify(top)}`
+        );
+        if (reps.length === 0) {
+          const ct = String((resp.headers as any)['content-type'] || '');
+          log.warn(
+            `DASH manifest parsed 0 variants (content-type=${ct}). First 300 chars: ${text.slice(0, 300)}`
+          );
+        }
+      } else {
+        // HLS master: #EXT-X-STREAM-INF lines
+        const lines = text.split(/\r?\n/);
+        const entries: Array<{ bw?: number; res?: string; codecs?: string }> = [];
+        for (let i = 0; i < lines.length; i++) {
+          const line = lines[i];
+          if (line.startsWith("#EXT-X-STREAM-INF:")) {
+            const bw = Number((line.match(/BANDWIDTH=(\d+)/) || [])[1]);
+            const res = (line.match(/RESOLUTION=(\d+x\d+)/) || [])[1];
+            const codecs = (line.match(/CODECS="([^"]+)"/) || [])[1];
+            entries.push({ bw: Number.isNaN(bw) ? undefined : bw, res, codecs });
+          }
+        }
+        entries.sort((a, b) => (b.bw || 0) - (a.bw || 0));
+        const top = entries.slice(0, 5).map(e => ({
+          bandwidth_kbps: e.bw ? Math.round(e.bw / 1000) : undefined,
+          resolution: e.res,
+          codecs: e.codecs,
+        }));
+        log.info(
+          `HLS master playlist summary url=${manifestUrl} total=${entries.length} variants=${JSON.stringify(top)}`
+        );
+        if (entries.length === 0) {
+          const ct = String((resp.headers as any)['content-type'] || '');
+          log.warn(
+            `HLS manifest parsed 0 variants (content-type=${ct}). First 300 chars: ${text.slice(0, 300)}`
+          );
+        }
+      }
+    } catch (e) {
+      log.warn(
+        `Failed to fetch/parse ${kind.toUpperCase()} manifest for logging url=${manifestUrl} err=${e instanceof Error ? e.message : e}`
+      );
+    }
+  }
+
+  /** Probe a manifest and return its top bitrate/resolution. 
+   * Needs to be there after finishing feature to give frontend all infos about quality
+  */
+  private async probeManifest(
+    host: string,
+    id: string,
+    kind: "dash" | "hls"
+  ): Promise<{ kind: "dash" | "hls"; url: string; topKbps: number; topRes?: string } | null> {
+    const url = this.manifestUrl(host, id, kind);
+    try {
+      const resp = await this.api.get<string>(url, {
+        headers: {
+          Accept:
+            kind === "dash"
+              ? "application/dash+xml,text/plain;q=0.9,*/*;q=0.8"
+              : "application/vnd.apple.mpegurl,application/x-mpegURL,text/plain;q=0.9,*/*;q=0.8",
+        },
+        responseType: "text",
+        transformResponse: [(r) => r],
+      });
+      const text = typeof resp.data === "string" ? resp.data : String(resp.data);
+      if (kind === "dash") {
+        const tags = text.match(/<Representation\b[^>]*>/g) ?? [];
```

Comment:
I don't like the excessive regexing here. We already pull in `m3u8-parser` and `@liveinstantly/dash-mpd-parser` to parse these.